internal/runners/ppm/convert.go

```diff
@@ -0,0 +1,230 @@
+package ppm
+
+import (
+	"os"
+	"os/signal"
+	"strings"
+	"sync"
+
+	"github.com/ActiveState/cli/internal/analytics"
+	"github.com/ActiveState/cli/internal/constants"
+	"github.com/ActiveState/cli/internal/failures"
+	"github.com/ActiveState/cli/internal/locale"
+	"github.com/ActiveState/cli/internal/logging"
+	"github.com/ActiveState/cli/internal/output"
+	"github.com/skratchdot/open-golang/open"
+)
+
+// convertAnswerCreate is the answer that the user can choose if they accept to create a virtual environment.  It is re-used at several places.
+var convertAnswerCreate string = locale.Tl("ppm_convert_answer_create", "Create Virtual Runtime Environment")
+
+// analyticsEventFunc is used to send analytics event
+type analyticsEventFunc func(string, string, string)
+
+// surveySelectFunc displays a menu with options that the user can select
+type surveySelectFunc func(message string, choices []string, defaultResponse string) (string, *failures.Failure)
+
+type conversionFlow struct {
+	survey   surveySelectFunc
+	out      output.Outputer
+	openURI  func(string) error
+	visitedC chan string
+	once     sync.Once
+}
+
+// StartConversionFlowIfNecessary checks if the user is in a project directory.
+// If not, they are asked to create a project, and (in a wizard-kind-of way) informed about the consequences.
+func (p *Ppm) StartConversionFlowIfNecessary() error {
+	// start conversion flow only if we cannot find a project file
+	if p.project == nil {
+		c := make(chan os.Signal, 1)
```

Comment:
I'm concerned with the use of channels here, I don't see why this should be necessary. Is it to facilitate your tests? If so, those tests should really be integration tests, not unit tests.

Reply:
No, it is not for the tests.  It is for handling the interrupt which needs to be done in a go function listening to this `c` channel.

And then - to send the right analytics method in case an interrupt is happening - I have to also send the visited menu items via a channel...  It is quite awkward, but an integration test would really not help here, as I am mostly concerned about the correct flow and return values, but also that the analytics function is called with the correct arguments.  And especially the latter cannot be verified with integration tests...

Reply:
Ok. I'll wait till the full PR then, but suffice to say I'd like the main conversion flow to look a lot dumber than it currently does. I think we can keep analytics and signal interrupts mostly separate from the rest of the conversion flow.